Step5 - 리팩터링

wefu0000 opened the pull request:
안녕하세요!
mvp  패턴 구현해 보았습니다.
이전 리뷰 주신거에 대해서 궁금한게 있어 댓글남겼습니다.
많은 조언 부탁 드립니다.

wefu00 committed:
fix: ArrayDeque 불변 타입으로 변경

wefu00 committed:
fix: operand 추가

wefu00 committed:
fix: ExpressionManager 네이밍 변경

wefu00 committed:
test: expression unit 테스트 추가

wefu00 committed:
build: mockk 추가

wefu00 committed:
feat: mvp 패턴구현

wefu00 committed:
test: presenter 테스트 추가

namjackson commented on domain/src/main/java/camp/nextstep/edu/calculator/domain/ExpressionItems.kt:
```diff
@@ -0,0 +1,42 @@
+package camp.nextstep.edu.calculator.domain
+
+class ExpressionItems {
+    private var expressionStack: List<ExpressionItem> = listOf()
```
> 불변 객체로 만들었는데요 private var expressionStack: List = listOf()
> 코멘트 주신 의견에 맞게 변경한게 맞는지 고민이 되서 문의 드려요

불변 객체는 내부 상태가 변경되지 않는 객체입니다! 
하지만 위의 expressionStack는 가변변수에 담겨져 있기 때문에, 가변 객체로 보기는 힘들거 같아요!
```
class ExpressionItems (val expressionStack: List<ExpressionItem>) {
    fun addExpression(expression: ExpressionItem) : ExpressionItems ... 
```
위 처럼 불변 컬렉션으로 생성해보면 어떨까요?

namjackson replied:
생성 시점에 list값을 주입받을수 있다면, 
관련 테스트 코드 작성시 좀더 유연하게 처리할수도 있을거같아요! 
(given절이 짧아질거같아요!)

wefu0000 replied:
네 의견 주신데로 변경하다보니
ExpressionPresenter 에 가변 ExpressionItems이 생기기 됐는데 괜찮을까요?
` private var expressionItems = ExpressionItems(emptyList())`
좋은 의견 감사합니다.

namjackson commented on app/src/main/java/camp/nextstep/edu/calculator/MainActivity.kt:
```diff
@@ -1,23 +1,25 @@
-class MainActivity : AppCompatActivity() {
     private lateinit var binding: ActivityMainBinding
-    private lateinit var expressionTextManager: ExpressionTextManager
+    private lateinit var expressionTextManager: Expression
```
이제는 MainActivity에서는 불필요한거 같아요!

wefu0000 replied:
제거 했습니다. 감사합니다

namjackson commented on app/src/main/java/camp/nextstep/edu/calculator/MainActivity.kt:
```diff
@@ -39,22 +41,27 @@ class MainActivity : AppCompatActivity() {
-            }
-        }
+        presenter.calculate()
+//        with(binding.textView) {
+//            val equalsText = expressionTextManager.calculate()
+//            if (equalsText.equals(text.toString(), false)) {
+//                Toast.makeText(this@MainActivity, "완성되지 않은 수식입니다", Toast.LENGTH_SHORT).show()
+//            } else {
+//                text = equalsText
+//            }
+//        }
```
불필요한 코드는 제거해주세요! 
(깃 로그에 남으니, 과감히 지워주셔도 좋아요!)

wefu0000 replied:
제거 했습니다. 감사합니다.

namjackson commented on app/src/test/java/camp/nextstep/edu/calculator/ExpressionPresenterTest.kt:
```diff
@@ -0,0 +1,175 @@
+    fun `숫자가 입력되면 수식에 추가되고 변경된 수식을 보여줘야 한다`() {
+        // given
+        val expressionSlot = slot<String>()
+        every { view.display(capture(expressionSlot)) } answers { nothing }
```
```suggestion
        every { view.display(capture(expressionSlot)) } just Runs
```
으로도 표현이 가능합니다!

wefu0000 replied:
그렇네요! 변경했습니다. 좋은의견 감사합니다.

namjackson commented on app/src/test/java/camp/nextstep/edu/calculator/ExpressionPresenterTest.kt:
```diff
@@ -0,0 +1,175 @@
+        // then
+        val actual: String = expressionSlot.captured
+        assertThat(actual).isEqualTo("1")
+        verify { view.display(actual) }
```
사실 slot을 사용하지않더라도,
verify내에서 특정 파라미터로 실행되었는지 확인할수 있어요! 😄

wefu0000 replied:
그렇네요! 변경했습니다. 좋은의견 감사합니다.

namjackson commented on app/src/main/java/camp/nextstep/edu/calculator/ExpressionPresenter.kt:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    override fun calculate() {
+        view.display(expression.calculate())
```
calculate가 실패한 경우에는 따로 처리되지 않는걸까요?
(ex> 완성되지 않은 수식)

wefu0000 replied:
완성되지 않은 수식에 대해 로직 추가하였습니다.

namjackson commented on domain/src/main/java/camp/nextstep/edu/calculator/domain/Expression.kt:
```diff
@@ -1,26 +1,91 @@
 
-    class NumberExpression(private var number: Int) : Expression() {
-        override fun getText(): String = number.toString()
+class Expression(
```
(참고만해주셔도 됩니다)
Expression 객체의 책임을 한번더 고민해봐도 좋을거같아요!
Expression 객체에는 입력값에 대한 처리, ExpressionItems 관리, calculate까지 해주고 있어요,
사실상 StringCaculator라고 봐꿔도 문제가 없을거 같단 생각이 들어요! 
Expression내의 expression관련 기능은 ExpressionItems에게 조금더 위임해봐도 좋을거 같단 생각이 들어요!

wefu0000 replied:
ExpressionItems 에 calculate 기능을 위임했습니다.
조금 더 좋은 구조로 변경이 가능하다면 더 많은 조언 부탁드립니다!

namjackson reviewed (changes requested):
안녕하세요 상인님!
mvp패턴 잘 적용해주셨네요! 👍 
몇가지 추가적인 코멘트 남겼으니, 확인해주세요!

wefu00 committed:
fix: 불 필요한 Expression 클래스 사용 제거

wefu00 committed:
test: 리뷰 내용 반영

1. answers -> just Runs 변경
2. 입력 동작이 하나일 경우 slot을 사용하지 않게 수정

wefu00 committed:
fix: 리뷰 내용 반영

1. ExpressionItems 생성자에서 List를 받고 ExpressionItems 반환하게 수정
2. Operand, Operation: Expression 하위로 구조 변경
3. ExpressionItems 에서 계산 수행

wefu00 committed:
fix: 리뷰 내용 반영

1. calculate 처리 실패 로직 추가

namjackson commented on domain/src/main/java/camp/nextstep/edu/calculator/domain/Expression.kt:
```diff
@@ -1,26 +1,71 @@
-    class NumberExpression(private var number: Int) : Expression() {
-        override fun getText(): String = number.toString()
+class Expression(private val inputTextConvertor: InputTextConvertor) {
+    fun addExpression(expressionItems: ExpressionItems, addText: String): ExpressionItems {
```
구현에는 정답이 없으니, 이부분은 마지막까지 고민해주시면 좋을거같아요!
상인님이 생각하시는 Expression, ExpressionItems, Caculator  각각의 책임은 무엇이라고 생각하시나요?
각각의 책임과 의존성이 섞여있는거 같아요!
ExpressionItems내의 Caculator에 대한 의존성이 있고,
Expression내에서 ExpressionItems 의존성이 있지만,
ExpressionItems의 관리는 Presenter에서 해주고 있는거 같아요,
각 객체의 관계를 고민해보면서 정리하면 좋을거 같아요

작성된 소스코드 기반으로 제가 생각했던 책임은 
ExpressionItems 표현식내의 상태를 관리하는 책임
Caculator은 문자열 기반을 계산하는 책임

그렇다면 Expression은 어떤 책임이 있을까요?
현재의 Expression객체는 입력에 대한 convertor와 표현식의 상태관리를 해주고 있어요
하지만 "표현식의 상태관리" ExpressionItems가 가져야할 책임은 아닐까요?

> ExpressionItems 에 calculate 기능을 위임했습니다.
> 조금 더 좋은 구조로 변경이 가능하다면 더 많은 조언 부탁드립니다!
ExpressionItems객체 또한 표현식 관리 책임 이외의 calculator 함수까지 있는 부분이 어색한거 같아요,
저의 커뮤니케이션에 미스가 있엇던거 같아요ㅠㅠ! 위임에 대한 부분은 표현식관리에 대한 부분이였어요! 
이부분은 다른 코멘트에서 추가적으로 남기겠습니다.

Expression 객체의 책임을 한번더 고민해보아요,
필요없다면 제거해주시는것도 좋고,
적당한 책임이 있다면, 해당 책임에 맞는 네이밍과 함수가 있으면 좋을거 같아요!
(ExpressionCaculator 와 같은 네이밍의 책임을 가져봐도 좋을거 같아요)

wefu0000 replied:
@namjackson 
구현에만 집중하다보니 역할 ,책임, 협력에 대해서 좀 미흡하게 생각했던거 같습니다.
ExpressionItems 에서 수식 관리에 대한 역할을 할 수 있게 위임하고 기능을 추가했습니다.
Expression은 제거 하였습니다.
계산은 Calculator 클래스에게 바로 요청하게 수정하였습니다.

조금 더 좋은 구조나 방법이 있다면 많은 조언 부탁 드립니다.

namjackson commented on domain/src/main/java/camp/nextstep/edu/calculator/domain/Expression.kt:
```diff
@@ -1,26 +1,71 @@
+    fun addExpression(expressionItems: ExpressionItems, addText: String): ExpressionItems {
+        return runCatching {
+            if (expressionItems.isEmpty()) addFirstExpression(expressionItems, addText)
+            else addOperandOrOperationText(expressionItems, addText)
```
expressionItems가 비었는지 확인하고, expressionItems의 함수(addFirstExpression, addOperandOrOperationText)를 실행시키기 보다, 
해당 기능들을 모두 위임해보면 어떨까요?
객체를 사용하기 보다, 객체에게 메세지를 던저서 객체가 일을 하게 위임해보면 좋을거 같아요!

namjackson commented on domain/src/main/java/camp/nextstep/edu/calculator/domain/Expression.kt:
```diff
@@ -1,26 +1,71 @@
+            addOperand(expressionItems, expression, addText)
+        } else {
+            addOperation(expressionItems, addText)
         }
```
이부분도 위코멘트와 동일한거 같아요 
expressionItems의 마지막 값이 무엇인지 가져와서, 확인하여, expressionItems의 함수를 실행시키기보단,
expressionItems에게 메세지(함수,파라미터 등의 방법)를 전달하여, 해당 동작을 수행하게끔 동작하면 어떨까요?

(아래의 비슷한 구조의 Expression 함수 모두 동일합니다)

namjackson commented on domain/src/main/java/camp/nextstep/edu/calculator/domain/ExpressionItem.kt:
```diff
@@ -0,0 +1,55 @@
+
+        object Division : Operation("/") {
+            override fun evaluate(x: Int, y: Int): Int = x / y
+            override fun getText(): String = " $operationText "
```
구분자가 " "으로 구분되는건 변경가능한 요구사항인거 같아요,
ExpressionItems 내에서 joinToString(" ")을 활용하는게 어떤가요?

wefu0000 replied:
네 좋습니다!
joinToString 사용하게 수정하였습니다

namjackson commented on domain/src/main/java/camp/nextstep/edu/calculator/domain/ExpressionItems.kt:
```diff
@@ -0,0 +1,42 @@
+
+class ExpressionItems(private val expressionStack: List<ExpressionItem>) {
+
+    fun addExpression(expression: ExpressionItem): ExpressionItems {
```
operator fun 같은 키워드도 적용해보면 좋을거같아요!

wefu0000 replied:
네 좋습니다!
operator plus 확장 함수 사용하게 수정하였습니다.

namjackson commented on domain/src/main/java/camp/nextstep/edu/calculator/domain/ExpressionItem.kt:
```diff
@@ -0,0 +1,55 @@
+sealed class ExpressionItem {
+    abstract fun getText(): String
+
+    class Operand(private var number: Int = 0) : ExpressionItem() {
```
Operand내의 상태를 불변으로 관리해보면 어떨까요?
불변 객체의 장점이 많은거 같아요 😄

wefu0000 replied:
네 맞습니다!
val 로 변경하였습니다.

namjackson reviewed (changes requested):
안녕하세요 상인님!
피드백 고생하셨습니다!
마지막 단계이니 조금더 고민해보면좋을만한
몇가지 추가적인 코멘트 남겼으니 확인해주세요!

namjackson commented:
혹시 이해안가는 부분이나, 질문있으시면, 태그로 코멘트 주셔도 되고, 슬랙 dm주셔도 됩니다!! 😄

wefu00 committed:
fix: 계산 기능은 Calculator 클래스 이용

wefu00 committed:
fix: Expression 클래스 제거

1. 수식관리에 필요한 기능은 ExpressionItems 에서 수행
2. ExpressionItems 에 Operator 확장함수 구현
3. Operand 인자값 불변으로 변경

wefu00 committed:
fix: joinToString 사용

namjackson commented on domain/src/test/java/camp/nextstep/edu/calculator/domain/ExpressionItemsTest.kt:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Test
+    fun `연산자 + 을추가하고 반환값이 빈문자인지 확인`() {
```
잘짜여진 테스트 케이스 명의 경우에는 
요구 사항 명세서로 불리기도 합니다!

> 입력된 피연산자가 없을 때, 연산자 - 를 추가하면, 아무런 변화가 없어야 한다

저는 테스트 케이스명에도 위처럼 ","를 활용해 준비,실행,검증 나눠서 작성하곤합니다

상인님만의 테스트케이스명 규칙을 지정해봐도 좋을거같아요!

namjackson commented on domain/src/test/java/camp/nextstep/edu/calculator/domain/ExpressionItemsTest.kt:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Test
+    fun `피연산자 1을추가하고 연산자 +을추가하고 피연산자 2를추가하고 삭제를 호출하여 반환값이 1 + 인지 확인`() {
+        expressionItems = expressionItems + Operand(1) + Operation.Addition + Operand(2)
+
+        val actual = expressionItems.removeLastExpression().getText()
+
+        assertThat(actual).isEqualTo("1 +")
+    }
```
개인적으로는 테스트 케이스에서 given, when, then 절을 활용하는 편입니다!  
준비, 실행, 검증 부를 나눠서 작성하면, 무엇을 테스트하는지 조금도 명확해지는거 같단 생각이 들어요!

```suggestion
    fun `1 + 2의 표현식이 주어졌을때, 삭제하면, 반환값이 1 + 이여야한다`() {
        // given : 1 + 2
        expressionItems = ExpressionItems(listOf(Operand(1), Operation.Addition, Operand(2)))

        // when : removeLastExpression
        val actual = expressionItems.removeLastExpression()
        
        // then : 1 +
        assertThat(actual.getText()).isEqualTo("1 +")
    }
```

namjackson reviewed (approved):
안녕하세요 상인님! 
피드백 반영 고생하셨습니다 👍 
반영 잘해주셨네요, 몇가지 참고하실만한 포인트들 코멘트로 남겼으니, 확인해주세요
추가로 코멘트 드릴부분은 없는거 같아서 
미션은 여기서 종료하도록하겠습니다! 고생하셨습니다 😄

namjackson commented:
MVP 미션 너무 고생많으셨습니다 😄 
미션 과정이 의미있는 시간이 되셨길 바랍니당!!
다음 미션들도 응원하겠습니다! 꼭 완주해보아요! 💪
